os/drivers/wifi_csi/wifi_csi.c

```diff
@@ -0,0 +1,183 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+#include <tinyara/wifi_csi/wifi_csi.h>
+#include <tinyara/wifi_csi/wifi_csi_common.h>
+#include <sys/types.h>
+#include <tinyara/wifi/rtk/wifi_structures.h>
+#include <debug.h>
+#include <errno.h>
+
+/**
+ * TODOs 
+ * - To implement the callback function in wifi_csi.c.
+ * - The callback function will share data to CSI Platform using mq_send.
+ * - next Phases Plan after this release
+ *		Phase-1. Remove dependency from RTK headers. Separate local headers for config params to only share between driver, CSI Platform
+ *		Phase-2: Multiple chip handling. Refactor implementation to upper and lower half as Suggested by Tj
+*/
+
+typedef void (*CSIRawDataReadyListener)(char *buf, int buf_len, int flags, void *userdata);
+
+static int csi_open(FAR struct file *filep);
+static int csi_close(FAR struct file *filep);
+static ssize_t csi_read(FAR struct file *filep, FAR char *buffer, size_t buflen);
+static int csi_ioctl(FAR struct file *filep, int cmd, unsigned long arg);
+
+static const struct file_operations g_csiops = {
+	csi_open,					/* open */
+	csi_close,					/* close */
+	csi_read,					/* read */
+	0,							/* write */
+	0,							/* seek */
+	csi_ioctl					/* ioctl */
+#ifndef CONFIG_DISABLE_POLL
+	, 0							/* poll */
+#endif
+};
+
+static rtw_csi_action_parm_t g_act_param __attribute__((aligned(64)))= {0,};
+
+int wifi_csi_register(FAR const char *name)
+{
+	csidbg("Registering csi_driver %s\n", CONFIG_WIFI_CSI_DEVPATH);
+	return register_driver(CONFIG_WIFI_CSI_DEVPATH, &g_csiops, 0666, NULL);
+}
+
+static int csi_open(FAR struct file *filep)
+{
+	csivdbg("DRIVER OPEN\n");
+	return 0;
+}
+
+static int csi_close(FAR struct file *filep)
+{
+	csivdbg("DRIVER CLOSE\n");
+	return 0;
+}
+
+static ssize_t csi_read(FAR struct file *filep, FAR char *buffer, size_t buflen)
+{
+	int len;
+	if (wifi_csi_report(buflen, buffer, &len) == -1) {
+		csidbg("ERROR: Wifi CSI report api\n");
+		return -EIO;
+	}
+	return len;
+}
+
+static int csi_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	CSIRawDataReadyListener user_cb;
+	int ret = 0;
+	switch (cmd) {
+	case CSIIOC_PARAM_SETZERO:
+		csivdbg("CSIIOC_PARAM_SETZERO\n");
+		memset(&g_act_param, 0, sizeof(rtw_csi_action_parm_t));
+		break;
+
+	case CSIIOC_SET_CONFIG:
+		csivdbg("CSIIOC_SET_CONFIG\n");
+		if (!arg) {
+			csidbg("ERROR: Invalid config arg\n");
+			ret = -EINVAL;
+			break;
+		}
+		rtw_csi_action_parm_t *g_act_param_t = ((rtw_csi_action_parm_t*)(arg));
```

Comment:
You should define your dev structure, it will be stored in private area of iNode.
I think this also can be added to dev structure.

Reply:
Hello Mr. Taejun,
the asked changes are already being implemented in a parallel task and I would soon publish the latest driver.

The current driver is a makeshift driver too fulfil requirements from Samsung Research America.